enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,427 @@
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "managed-bootimages-platform-check"
+spec:
+  failurePolicy: Fail
+  paramKind:
+    apiVersion: config.openshift.io/v1
+    kind: Infrastructure
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["operator"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["MachineConfiguration"]
+  validations:
+    - expression: "has(object.spec.MachineBootImageConfig) && param.status.platformStatus.Type != `GCP`"
+      message: "This feature is only supported on these platforms: GCP"
+```
+This would need an accompanying binding:
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: "managed-bootimages-platform-check-binding"
+spec:
+  policyName: "managed-bootimages-platform-check"
+  validationActions: [Deny]
+  paramRef:
+    name: "cluster"
+    namespace: "default"
+```
+#### Tracking boot image history
+
+This proposal will also introduce a new CR, `MachineSetBootImageHistory` for tracking boot image history. As a starting point, here is a stub type definition for this:
+
+```
+type MachineSetBootImageHistory struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MachineSetBootImageHistorySpec   `json:"spec,omitempty"`
+	Status MachineSetBootImageHistoryStatus `json:"status,omitempty"`
+}
+
+// MachineSetBootImageHistorySpec defines the desired state of MachineSetBootImageHistory
+type MachineSetBootImageHistorySpec struct {
+	MachineSetName string                   `json:"machineSetName"`
+	Details        []BootImageHistoryDetail `json:"details"`
+}
+
+// MachineSetBootImageHistoryStatus defines the observed state of MachineSetBootImageHistory
+type MachineSetBootImageHistoryStatus struct {
+}
+
+// BootImageHistoryDetail is the struct for each element in the Details array
+type BootImageHistoryDetail struct {
+	Index          int         `json:"index"`
+	UpdatedTime    metav1.Time `json:"updatedTime"`
+	BootImageRef   string      `json:"bootImageRef"`
+}
+
+// MachineSetBootImageHistoryList contains a list of MachineSetBootImageHistory
+type MachineSetBootImageHistoryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []MachineSetBootImageHistory `json:"items"`
+}
+```
+There will be one instance of this per `Machineset`. It will be updated by the MSBIC as `Machinesets` are created/updated and will exist in the same namespace as the `MachineSet`. This CRD will also need to support MAPI and CAPI backed `MachineSets`. The goal of this is to provide information about the "lineage" of a `MachineSet` to the user. The user can then manually restore their `MachineSet` to an earlier state if they wish to do so by following documentation. 
```

Comment:
Can you explain the naming convention for these resources please and how they will support both Machine API and Cluster API simultaneously

Reply:
As you mentioned `MachineDeployments` in another comment, I gave this approach another think. We don't necessarily want to track the history of a `MachineSet` necessarily, but really the boot image history of a machine resource. So I simplified the overall CRD name, and added a `name`, `apigroup`, and `kind` to the reference type. The `name` field might be a little overkill, as they will always match the name of the CR by design. I also added some YAML examples of what I envision these CRs will look like.

I do want to note that this won't be part of the MVP, so by the time we get to this, we may be in a Cluster API only world...but probably still worth planning for all the scenarios (:

Reply:
Boot images won't change for individual machine resources, they are immutable, so in that case, I'm not sure I understand your intention above.

Reply:
Ah sorry, I can see how my wording in there can be confusing. Instead of a machine resource, I should have said a machine management resource, such as a machineset(could be MAPI or CAPI), a MachineDeployment, or perhaps something else that we will introduce in the future. I added a couple of YAML examples of what I think it would look like. Does that make more sense?